File: Jadwal/Packages/JadwalCore/Sources/Domain/ScheduleTimeline.swift
import Foundation

public enum ClassLiveStatus: Equatable, Sendable {
    case completed
    case inProgress
    case startingSoon(minutes: Int)
    case upcoming
    case cancelled
    case changed(summary: String)

    public var isEmphasized: Bool {
        switch self {
        case .inProgress, .startingSoon:
            return true
        default:
            return false
        }
    }

    public static func compute(for period: PeriodOccurrence, currentTime: Date = Date()) -> ClassLiveStatus {
        if let change = period.changeRecord, !change.isAcknowledged {
            if change.changeType == .removed {
                return .cancelled
            }
            return .changed(summary: change.summaryMessage)
        }

        let calendar = Calendar.current
        let currentHour = calendar.component(.hour, from: currentTime)
        let currentMinute = calendar.component(.minute, from: currentTime)
        let currentMinutes = currentHour * 60 + currentMinute

        guard let startMinutes = parseMinutes(from: period.startTime),
              let endMinutes = parseMinutes(from: period.endTime) else {
            return .upcoming
        }

        if currentMinutes > endMinutes {
            return .completed
        } else if currentMinutes >= startMinutes && currentMinutes <= endMinutes {
            return .inProgress
        } else {
            let diff = startMinutes - currentMinutes
            if diff > 0 && diff <= 15 {
                return .startingSoon(minutes: diff)
            } else {
                return .upcoming
            }
        }
    }

    private static func parseMinutes(from timeStr: String) -> Int? {
        let parts = timeStr.trimmingCharacters(in: .whitespaces).split(separator: ":")
        guard parts.count == 2,
              let h = Int(parts[0]),
              let m = Int(parts[1]) else {
            return nil
        }
        return h * 60 + m
    }
}

public enum ScheduleTimelineItem: Identifiable, Sendable {
    case classPeriod(period: PeriodOccurrence, status: ClassLiveStatus)
    case breakBlock(id: String, name: String, startTime: String, endTime: String, durationMinutes: Int)

    public var id: String {
        switch self {
        case .classPeriod(let period, _):
            return period.id
        case .breakBlock(let id, _, _, _, _):
            return id
        }
    }

    public var startTime: String {
        switch self {
        case .classPeriod(let period, _):
            return period.startTime
        case .breakBlock(_, _, let start, _, _):
            return start
        }
    }

    public var endTime: String {
        switch self {
        case .classPeriod(let period, _):
            return period.endTime
        case .breakBlock(_, _, _, let end, _):
            return end
        }
    }
}

public enum ScheduleTimelineBuilder {
    public static func parseMinutes(from timeStr: String) -> Int? {
        let parts = timeStr.trimmingCharacters(in: .whitespaces).split(separator: ":")
        guard parts.count == 2,
              let h = Int(parts[0]),
              let m = Int(parts[1]) else {
            return nil
        }
        return h * 60 + m
    }

    public static func formatDuration(minutes: Int) -> String {
        if minutes >= 60 {
            let hours = minutes / 60
            let remainder = minutes % 60
            if remainder == 0 {
                return "\(hours)h"
            }
            return "\(hours)h \(remainder)m"
        }
        return "\(minutes)m"
    }

    public static func buildTimeline(from periods: [PeriodOccurrence], currentTime: Date = Date()) -> [ScheduleTimelineItem] {
        guard !periods.isEmpty else { return [] }

        // Sort chronologically by startTime
        let sorted = periods.sorted { (p1, p2) -> Bool in
            let m1 = parseMinutes(from: p1.startTime) ?? 0
            let m2 = parseMinutes(from: p2.startTime) ?? 0
            return m1 < m2
        }

        var items: [ScheduleTimelineItem] = []

        for i in 0..<sorted.count {
            let current = sorted[i]
            let status = ClassLiveStatus.compute(for: current, currentTime: currentTime)
            items.append(.classPeriod(period: current, status: status))

            // Check if there is a gap to next period
            if i + 1 < sorted.count {
                let next = sorted[i + 1]
                if let endMin = parseMinutes(from: current.endTime),
                   let nextStartMin = parseMinutes(from: next.startTime) {
                    let gap = nextStartMin - endMin
                    if gap >= 10 {
                        let name = intelligentBreakName(startMin: endMin, endMin: nextStartMin, durationMin: gap)
                        let breakId = "break_\(current.endTime)_\(next.startTime)"
                        items.append(.breakBlock(
                            id: breakId,
                            name: name,
                            startTime: current.endTime,
                            endTime: next.startTime,
                            durationMinutes: gap
                        ))
                    }
                }
            }
        }

        return items
    }

    private static func intelligentBreakName(startMin: Int, endMin: Int, durationMin: Int) -> String {
        // e.g. 07:00 to 08:50 (Morning prep)
        if startMin < 8 * 60 {
            return "Morning Preparation"
        }
        // e.g. 10:00 to 11:30 (Recess)
        if startMin >= 10 * 60 && startMin < 12 * 60 {
            return "Recess"
        }
        // e.g. 12:00 to 14:00 (Lunch & Namaz)
        if startMin >= 12 * 60 && startMin < 14 * 60 {
            return "Lunch & Namaz Break"
        }
        // e.g. afternoon
        if startMin >= 14 * 60 {
            return "Afternoon Break"
        }
        return "Break"
    }

    /// Splits items into two horizontal rows according to day schedule rules (Mon-Thu, Fri, Sat)
    public static func splitIntoTwoHorizontalRows(
        items: [ScheduleTimelineItem],
        day: DayOfWeek
    ) -> (row1: [ScheduleTimelineItem], row2: [ScheduleTimelineItem]) {
        guard !items.isEmpty else { return ([], []) }

        switch day {
        case .monday, .tuesday, .wednesday, .thursday, .friday:
            // Find the index of Period 5 to split cleanly
            if let p5Index = items.firstIndex(where: {
                if case .classPeriod(let p, _) = $0 {
                    return p.periodName.contains("5")
                }
                return false
            }) {
                let splitPoint = p5Index + 1
                let row1 = Array(items[0..<splitPoint])
                let row2 = Array(items[splitPoint..<items.count])
                return (row1, row2)
            }

        case .saturday:
            // On Saturday, Row 1 includes Periods 1–4 and the Recess break right after Period 4.
            // Row 2 begins with Period 5 and continues through Period 8.
            if let p4Index = items.firstIndex(where: {
                if case .classPeriod(let p, _) = $0 {
                    return p.periodName.contains("4")
                }
                return false
            }) {
                var splitPoint = p4Index + 1
                if splitPoint < items.count, case .breakBlock = items[splitPoint] {
                    splitPoint += 1
                }
                let row1 = Array(items[0..<splitPoint])
                let row2 = Array(items[splitPoint..<items.count])
                return (row1, row2)
            }

        case .sunday:
            return (items, [])
        }

        // Fallback: split near the middle
        let mid = (items.count + 1) / 2
        let row1 = Array(items[0..<mid])
        let row2 = Array(items[mid..<items.count])
        return (row1, row2)
    }

    /// Divides timeline items into two balanced columns (Morning / Afternoon)
    /// to ensure zero vertical scrolling on standard viewports.
    public static func splitIntoTwoColumns(items: [ScheduleTimelineItem]) -> (morning: [ScheduleTimelineItem], afternoon: [ScheduleTimelineItem]) {
        guard !items.isEmpty else { return ([], []) }

        // Try to split at a major break (Lunch & Namaz or around noon)
        if let lunchIndex = items.firstIndex(where: {
            if case .breakBlock(_, let name, _, _, _) = $0, name.localizedCaseInsensitiveContains("Lunch") {
                return true
            }
            return false
        }) {
            // Include morning up to and including lunch break or split right after lunch
            let splitPoint = lunchIndex + 1
            let morning = Array(items[0..<splitPoint])
            let afternoon = Array(items[splitPoint..<items.count])
            return (morning, afternoon)
        }

        // Fallback: split near the middle
        let mid = (items.count + 1) / 2
        let morning = Array(items[0..<mid])
        let afternoon = Array(items[mid..<items.count])
        return (morning, afternoon)
    }
}

public struct DayScheduleRule: Sendable, Equatable {
    public let day: DayOfWeek
    public let hasPhysicalTraining: Bool
    public let isHalfDay: Bool
    public let daySubtitle: String
    public let row1Title: String
    public let row2Title: String

    public static func rule(for day: DayOfWeek) -> DayScheduleRule {
        switch day {
        case .monday, .tuesday, .wednesday, .thursday:
            return DayScheduleRule(
                day: day,
                hasPhysicalTraining: true,
                isHalfDay: false,
                daySubtitle: "Full Academic Schedule • 10 Periods",
                row1Title: "Morning Prep & Early Sessions (Periods 1–5)",
                row2Title: "Midday & Afternoon Sessions (Periods 6–10)"
            )
        case .friday:
            return DayScheduleRule(
                day: day,
                hasPhysicalTraining: false,
                isHalfDay: false,
                daySubtitle: "Jumua Mubarak • No Morning PT (Periods 2–10)",
                row1Title: "Morning Sessions (Periods 2–5)",
                row2Title: "Midday & Post-Jumua Sessions (Periods 6–10)"
            )
        case .saturday:
            return DayScheduleRule(
                day: day,
                hasPhysicalTraining: false,
                isHalfDay: true,
                daySubtitle: "Saturday Half-Day • Concludes at 1:15 PM",
                row1Title: "Early Morning Sessions (Periods 1–4)",
                row2Title: "Late Morning Sessions (Periods 5–8)"
            )
        case .sunday:
            return DayScheduleRule(
                day: day,
                hasPhysicalTraining: false,
                isHalfDay: true,
                daySubtitle: "Weekend • No Classes Scheduled",
                row1Title: "Weekend",
                row2Title: ""
            )
        }
    }
}

File: src/Jadwal.Domain/Models/ScheduleTimeline.cs
using System.Text.Json.Serialization;
using Jadwal.Domain.Enums;

namespace Jadwal.Domain.Models;

public record BreakBlock(
    string Id,
    string Name,
    string StartTime,
    string EndTime,
    int DurationMinutes
);

public enum ScheduleTimelineItemKind
{
    ClassPeriod,
    BreakBlock
}

public record ScheduleTimelineItem
{
    public ScheduleTimelineItemKind Kind { get; init; }
    public PeriodOccurrence? Period { get; init; }
    public ClassLiveStatus Status { get; init; }
    public BreakBlock? Break { get; init; }

    public static ScheduleTimelineItem ForPeriod(PeriodOccurrence period, ClassLiveStatus status) => new()
    {
        Kind = ScheduleTimelineItemKind.ClassPeriod,
        Period = period,
        Status = status
    };

    public static ScheduleTimelineItem ForBreak(BreakBlock breakBlock) => new()
    {
        Kind = ScheduleTimelineItemKind.BreakBlock,
        Break = breakBlock,
        Status = ClassLiveStatus.Upcoming
    };
}

public record DayScheduleRule(
    JadwalDayOfWeek Day,
    bool HasPhysicalTraining,
    bool IsHalfDay,
    string DaySubtitle,
    string Row1Title,
    string Row2Title
)
{
    public static DayScheduleRule RuleFor(JadwalDayOfWeek day) => day switch
    {
        JadwalDayOfWeek.Monday or JadwalDayOfWeek.Tuesday or JadwalDayOfWeek.Wednesday or JadwalDayOfWeek.Thursday => new(
            Day: day,
            HasPhysicalTraining: true,
            IsHalfDay: false,
            DaySubtitle: "Full Academic Schedule • 10 Periods",
            Row1Title: "Morning Prep & Early Sessions (Periods 1–5)",
            Row2Title: "Midday & Afternoon Sessions (Periods 6–10)"
        ),
        JadwalDayOfWeek.Friday => new(
            Day: day,
            HasPhysicalTraining: false,
            IsHalfDay: false,
            DaySubtitle: "Jumua Mubarak • No Morning PT (Periods 2–10)",
            Row1Title: "Morning Sessions (Periods 2–5)",
            Row2Title: "Midday & Post-Jumua Sessions (Periods 6–10)"
        ),
        JadwalDayOfWeek.Saturday => new(
            Day: day,
            HasPhysicalTraining: false,
            IsHalfDay: true,
            DaySubtitle: "Saturday Half-Day • Concludes at 1:15 PM",
            Row1Title: "Early Morning Sessions (Periods 1–4)",
            Row2Title: "Late Morning Sessions (Periods 5–8)"
        ),
        JadwalDayOfWeek.Sunday => new(
            Day: day,
            HasPhysicalTraining: false,
            IsHalfDay: true,
            DaySubtitle: "Weekend • No Classes Scheduled",
            Row1Title: "Weekend",
            Row2Title: string.Empty
        ),
        _ => RuleFor(JadwalDayOfWeek.Monday)
    };
}

public static class ScheduleTimelineBuilder
{
    private static int? ParseMinutes(string? timeStr)
    {
        if (string.IsNullOrWhiteSpace(timeStr)) return null;
        var parts = timeStr.Trim().Split(':');
        if (parts.Length >= 2 && int.TryParse(parts[0], out var h) && int.TryParse(parts[1], out var m))
        {
            return h * 60 + m;
        }
        return null;
    }

    public static string FormatDuration(int minutes)
    {
        var h = minutes / 60;
        var m = minutes % 60;
        if (h > 0 && m > 0) return $"{h}h {m}m";
        if (h > 0) return $"{h}h";
        return $"{m}m";
    }

    private static string IntelligentBreakName(int startMin, int endMin, int durationMin)
    {
        if (startMin < 8 * 60) return "Morning Preparation";
        if (startMin >= 10 * 60 && startMin < 12 * 60) return "Recess";
        if (startMin >= 12 * 60 && startMin < 14 * 60) return "Lunch & Namaz Break";
        if (startMin >= 14 * 60) return "Afternoon Break";
        return "Break";
    }

    public static IReadOnlyList<ScheduleTimelineItem> BuildTimeline(
        IEnumerable<PeriodOccurrence> periods,
        TimeOnly? currentTime = null)
    {
        var now = currentTime ?? TimeOnly.FromDateTime(DateTime.Now);
        var sorted = periods
            .OrderBy(p => ParseMinutes(p.StartTime) ?? 0)
            .ToList();

        var items = new List<ScheduleTimelineItem>();

        for (var i = 0; i < sorted.Count; i++)
        {
            var current = sorted[i];
            var status = current.ChangeRecord != null
                ? ClassLiveStatus.Changed
                : ClassLiveStatus.Compute(current.StartTime, current.EndTime, now);

            items.Add(ScheduleTimelineItem.ForPeriod(current, status));

            if (i + 1 < sorted.Count)
            {
                var next = sorted[i + 1];
                var endMin = ParseMinutes(current.EndTime);
                var nextStartMin = ParseMinutes(next.StartTime);

                if (endMin.HasValue && nextStartMin.HasValue)
                {
                    var gap = nextStartMin.Value - endMin.Value;
                    if (gap >= 10)
                    {
                        var name = IntelligentBreakName(endMin.Value, nextStartMin.Value, gap);
                        var breakBlock = new BreakBlock(
                            Id: $"break_{current.EndTime}_{next.StartTime}",
                            Name: name,
                            StartTime: current.EndTime,
                            EndTime: next.StartTime,
                            DurationMinutes: gap
                        );
                        items.Add(ScheduleTimelineItem.ForBreak(breakBlock));
                    }
                }
            }
        }

        return items.AsReadOnly();
    }

    public static (IReadOnlyList<ScheduleTimelineItem> Row1, IReadOnlyList<ScheduleTimelineItem> Row2) SplitIntoTwoHorizontalRows(
        IReadOnlyList<ScheduleTimelineItem> items,
        JadwalDayOfWeek day)
    {
        if (items.Count == 0) return (Array.Empty<ScheduleTimelineItem>(), Array.Empty<ScheduleTimelineItem>());

        switch (day)
        {
            case JadwalDayOfWeek.Monday or JadwalDayOfWeek.Tuesday or JadwalDayOfWeek.Wednesday or JadwalDayOfWeek.Thursday or JadwalDayOfWeek.Friday:
            {
                // Find Period 5
                var p5Index = -1;
                for (var i = 0; i < items.Count; i++)
                {
                    if (items[i].Kind == ScheduleTimelineItemKind.ClassPeriod &&
                        items[i].Period?.PeriodName?.Contains("5") == true)
                    {
                        p5Index = i;
                        break;
                    }
                }

                if (p5Index >= 0)
                {
                    var splitPoint = p5Index + 1;
                    return (items.Take(splitPoint).ToList().AsReadOnly(), items.Skip(splitPoint).ToList().AsReadOnly());
                }
                break;
            }

            case JadwalDayOfWeek.Saturday:
            {
                // On Saturday: Row 1 contains Period 4 plus the Recess break after Period 4.
                // Row 2 begins with Period 5 and continues through Period 8.
                var p4Index = -1;
                for (var i = 0; i < items.Count; i++)
                {
                    if (items[i].Kind == ScheduleTimelineItemKind.ClassPeriod &&
                        items[i].Period?.PeriodName?.Contains("4") == true)
                    {
                        p4Index = i;
                        break;
                    }
                }

                if (p4Index >= 0)
                {
                    var splitPoint = p4Index + 1;
                    if (splitPoint < items.Count && items[splitPoint].Kind == ScheduleTimelineItemKind.BreakBlock)
                    {
                        splitPoint++;
                    }
                    return (items.Take(splitPoint).ToList().AsReadOnly(), items.Skip(splitPoint).ToList().AsReadOnly());
                }
                break;
            }

            case JadwalDayOfWeek.Sunday:
                return (items, Array.Empty<ScheduleTimelineItem>());
        }

        // Fallback: split in middle
        var mid = (items.Count + 1) / 2;
        return (items.Take(mid).ToList().AsReadOnly(), items.Skip(mid).ToList().AsReadOnly());
    }
}
